compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -1115,8 +1115,11 @@ class Typer(@constructorOnly nestingLevel: Int = 0) extends Namer
       fullyDefinedType(tree.tpe, "block", tree.srcPos)
       var avoidingType = TypeOps.avoid(tree.tpe, localSyms)
       val ptDefined = isFullyDefined(pt, ForceDegree.none)
-      if (ptDefined && !(avoidingType.widenExpr <:< pt)) avoidingType = pt
-      val tree1 = ascribeType(tree, avoidingType)
+      var tree1 = ascribeType(tree, avoidingType)
```

Comment:
I liked the old logic better: First figure out what type to ascribe, then do it.